object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java

```diff
@@ -428,26 +561,33 @@ private void handleSuccessorBlocks(
     }
     List<Node> nodes = block.getNodes();
     for (Block succ : getRelevantSuccessors(block)) {
-      Set<LocalVarWithTree> defsCopy = new HashSet<>(defs);
-      Set<LocalVarWithTree> toRemove = new HashSet<>();
+      Set<Set<LocalVarWithTree>> defsCopy = new LinkedHashSet<>(defs);
+      Set<Set<LocalVarWithTree>> toRemove = new LinkedHashSet<>();
 
       CFStore succRegularStore = analysis.getInput(succ).getRegularStore();
-      for (LocalVarWithTree assign : defs) {
-
+      for (Set<LocalVarWithTree> setAssign : defs) {
+        Set<LocalVarWithTree> setAssignCopy = new LinkedHashSet<>(setAssign);
         // If the successor block is the exit block or if the variable is going out of scope
         if (succ instanceof SpecialBlockImpl
-            || succRegularStore.getValue(assign.localVar) == null) {
+            || setAssignCopy.stream()
+                .allMatch(assign -> succRegularStore.getValue(assign.localVar) == null)) {
           if (nodes.size() == 0) { // If the cur block is special or conditional block
-            checkMustCall(assign, succRegularStore, outOfScopeReason);
+            checkMustCall(setAssignCopy, succRegularStore, outOfScopeReason);
 
           } else { // If the cur block is Exception/Regular block then it checks MustCall
             // annotation in the store right after the last node
             Node last = nodes.get(nodes.size() - 1);
             CFStore storeAfter = typeFactory.getStoreAfter(last);
-            checkMustCall(assign, storeAfter, outOfScopeReason);
+            checkMustCall(setAssignCopy, storeAfter, outOfScopeReason);
           }
 
-          toRemove.add(assign);
+          toRemove.add(setAssignCopy);
+        } else {
+          setAssignCopy.removeIf(assign -> succRegularStore.getValue(assign.localVar) == null);
```

Comment:
This needs a comment indicating you are handling the case where _some_ vars go out of scope in the set.  Also, I think you can just make `setAssignCopy` in this else case rather than earlier, since this is the only place where mutation occurs.